pkg/network/prober/prober.go

```diff
@@ -48,3 +50,36 @@ func Do(ctx context.Context, target, headerValue string) (int, string, error) {
 	}
 	return resp.StatusCode, string(body), nil
 }
+
+// Done is a function that is executed after each probe returns true if the
+// probe should be finished, or false to probe again.
+type Done func(status int, body string, err error) bool
+
+// Default settings for out exponential backoff.
+var backOffSettings = wait.Backoff{
+	Duration: 50 * time.Millisecond,
+	Factor:   1.3,
+	Jitter:   0.2,
+	Steps:    22,
+}
+
+// DoAsync starts a go routine that executes `Do` using K8s Wait library.
+// After each call to `Do` the passed in `done` callback is executed that
+// should return true to finish probing or false to continue.
+// If no response is ever received or terminal condition is not reached
+// `Done` is invoked with (0, "", wait.ErrWaitTimeout).
+func DoAsync(ctx context.Context, target, headerValue string, done Done) {
+	doAsync(ctx, target, headerValue, done, backOffSettings)
+}
+
+func doAsync(ctx context.Context, target, headerValue string, done Done, bo wait.Backoff) {
+	go func() {
+		// `err` is always `wait.ErrWaitTimeout`
+		if err := wait.ExponentialBackoff(bo, func() (bool, error) {
+			st, body, err := Do(ctx, target, headerValue)
+			return done(st, body, err), nil
+		}); err != nil {
+			done(0, "", err)
+		}
+	}()
```

Comment:
I'm a little worried about leaking Go routines.  Also, do you think we should either check the `ctx`'s done channel or pass it through so that the `Done` callback can?

Reply:
- Why would they leak? They are time bounded.
- `Do` does request with context, which means, after some point all retries will immediately fail with DeadlineExceeded, if the context was ever bound to a timeout.
 I guess we can optimize for that case and stop iterating early.

Reply:
As discussed in chat, it's a bounded leak, but what I'm worried about is subsequent reconciles (perhaps several) that may occur in parallel to the first async probe.  As you put it: I think we need a "manager" concept. to handle deduplicating the async probes.

Reply:
Done.